### DISCUSSION TITLE
Differentiate between action and filter hooks

### DISCUSSION THREAD

**Discussion by smilledge** [directus/directus#7843] [Feature Requests]
Currently the Directus hook system is a bit confusing for a few reasons:
1. It's not clear when you can return a value to change a process and when you can only listen to an action that is occurring
2. The logic for returned values isn't clear (defined inline for each event is emitted)
3. There's currently no way to control the order of hooks

Would it make sense to have two different hook types? Similar to [actions and filters](https://developer.wordpress.org/plugins/hooks/#actions-vs-filters) in Wordpress:

- an **action** takes the info it receives, does something with it, and returns nothing. In other words: it acts on something and then exits, returning nothing back to the calling hook.
- a **filter** takes the info it receives, modifies it somehow, and returns it. In other words: it filters something and passes it back to the hook for further use.

**Comment by rijkvanzanten** [comment]
This makes total sense to me 👍🏻 

@nickrum thoughts? (cc @benhaynes)

**Reply by benhaynes** [reply]
Maybe I'm misunderstanding, but couldn't we always use `before` as "filters"? If you catch an event before it happens, then you should be able to edit that event traffic... and all the normal (after) events would simply be triggers for external logic (actions).

What am I missing here? Is there unclear overlap?

**Reply by rijkvanzanten** [reply]
Yeah that's exactly the case. It's a difference in documenting / naming. So instead of calling it all "hooks", we differentiate between "filter" and "action" hooks 👍🏻

**Comment by nickrum** [comment]
This has been done before the v9 release. Thanks again for the suggestion 👍
